src/inspector/inspector_socket_server.cpp

```diff
@@ -338,7 +338,11 @@ bool InspectorSocketServer::HasTargets() {
 
 bool InspectorSocketServer::Start() {
   state_ = ServerState::kRunning;
-  std::thread([this]() {
+
+  // It feels safer to lock before the thread jump; if we get through Start, then Stop, then the thread gets scheduled now touching deleted memory that's not good either.
+  // In real-life this won't really matter; it's only an issue in unit tests that quickly bring up and tear down runtimes.
+  std::unique_lock<std::mutex> tcp_server_lock(mutex_tcp_server_);
+  std::thread([this, lock = std::move(tcp_server_lock)]() {
     tcp_server_ = std::make_shared<tcp_server>(

```

Comment:
>tcp_server_ [](http://example.com/codeflow?start=4&length=11)

It seems that we still have a race condition: the thread can start a little bit later than the Stop is executed.
It would be better to use the mutex to have any access to the tcp_server_ and state_ variables.
Also, we probably need a condition variable to notify code in the Stop method that the thread is finished.

Reply:
We lock in Start, then move the unique_lock to the thread, where it unlocks when the async callback is finished; Start / Stop must be called from the same thread. I think the change takes care of race access conditions on tcp_server_ but I agree we need to take another look at the implementation top-down.

Reply:
Locks are per thread. We must not move them to another thread. We just need to create a new lock inside of the thread when we access the tcp_server_ variable. Inside of the lambda we need to create a local variable for it to make sure that we can unlock the lock before we call run.

Reply:
If we could move the lock to another thread it would never be released because it would be released only when the lambda is destroyed and the the Stop would never able to acquire the mutex. The issue is not with locks, but with mutexes which are typically bound to a thread when locked, i.e. if a thread acquires a mutex, then other threads cannot acquire it and must wait.

Reply:
Thank you, that makes sense.